packages/message-relayer/src/service.ts

```diff
@@ -236,6 +245,25 @@ export class MessageRelayerService extends BaseService<MessageRelayerOptions> {
           await this._relayMessageToL1(message, proof)
         }
 
+        if (messages.length === 0) {
+          this.logger.info('Did not find any L2->L1 messages', {
+            retryAgainInS: Math.floor(this.options.pollingInterval / 1000),
+          })
+        } else {
+          // Clear the event cache to avoid keeping every single event in memory and eventually
+          // getting OOM killed. Messages are already sorted in ascending order so the last message
+          // will have the highest block number. We also know that we will already have dealt with
+          // all messages with a block number equal to or lower than this highest block number. So
+          // we now just need to keep any events with an even higher block number.
+          const lastMessage = messages[messages.length - 1]
+          this.state.eventCache = this.state.eventCache.filter((event) => {
+            return (
+              event.blockNumber >
+              lastMessage.parentTransactionIndex + this.options.l2BlockOffset
+            )
+          })
```

Comment:
Why do we need to filter out events vs just empty the entire `messages` list? What is the condition where a message will still be left in the list after going through all the messages in the above `for loop`?

Definitely could be missing some logic but was getting tripped up there

Reply:
We maintain a cache of events: https://github.com/ethereum-optimism/optimism/blob/c88004371a2aff7b642c8d2aa2a44e925336eafe/packages/message-relayer/src/service.ts#L282

This speeds up the process of looking for `StateBatchAppended` events: https://github.com/ethereum-optimism/optimism/blob/c88004371a2aff7b642c8d2aa2a44e925336eafe/packages/message-relayer/src/service.ts#L276-L280

However, we don't need these events anymore once we've processed the messages that correspond to . So we can drop them from the cache.